RecoLocalFastTime/FTLClusterizer/plugins/MTDClusterProducer.cc

```diff
@@ -0,0 +1,178 @@
+};
+
+
+//---------------------------------------------------------------------------
+//!  Constructor: set the ParameterSet and defer all thinking to setupClusterizer().
+//---------------------------------------------------------------------------
+MTDClusterProducer::MTDClusterProducer(edm::ParameterSet const& conf) 
+  : 
+  btlHits_(consumes< FTLRecHitCollection >( conf.getParameter<edm::InputTag>("srcBarrel"))),
+  etlHits_(consumes< FTLRecHitCollection >( conf.getParameter<edm::InputTag>("srcEndcap"))),
+  ftlbInstance_(conf.getParameter<std::string>("BarrelClusterName")),
+  ftleInstance_(conf.getParameter<std::string>("EndcapClusterName")),
+  clusterMode_( conf.getParameter<std::string>("ClusterMode") ),
+  clusterizer_(nullptr)          // the default, in case we fail to make one
+{
+
+  //--- Declare to the EDM what kind of collections we will be making.
+  produces<FTLClusterCollection>(ftlbInstance_); 
+  produces<FTLClusterCollection>(ftleInstance_); 
+  
+  //--- Make the algorithm(s) according to what the user specified
+  //--- in the ParameterSet.
+  if ( clusterMode_ == "MTDThresholdClusterizer" ) {
+    clusterizer_ = std::make_unique<MTDThresholdClusterizer>(conf);
+  } 
+  else {
+    throw cms::Exception("MTDClusterProducer") << "[MTDClusterProducer]:"
+					       <<" choice " << clusterMode_ << " is invalid.\n"
+					       << "Possible choices:\n" 
+					       << "    MTDThresholdClusterizer";
+  }
+}
+
+// Destructor
+MTDClusterProducer::~MTDClusterProducer() { 
+}  
+
+// Configuration descriptions
+void
+MTDClusterProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("srcBarrel", edm::InputTag("mtdRecHits:FTLBarrel"));
+  desc.add<edm::InputTag>("srcEndcap", edm::InputTag("mtdRecHits:FTLEndcap"));
+  desc.add<std::string>("BarrelClusterName", "FTLBarrel");
+  desc.add<std::string>("EndcapClusterName", "FTLEndcap");
+  desc.add<std::string>("ClusterMode","MTDThresholdClusterizer");
+  desc.add<double>("HitThreshold", 0.);
+  desc.add<double>("SeedThreshold", 0.);
+  desc.add<double>("ClusterThreshold", 0.);
+  descriptions.add("mtdClusterProducer", desc);
+}
+  
+//---------------------------------------------------------------------------
+//! The "Event" entrypoint: gets called by framework for every event
+//---------------------------------------------------------------------------
+void MTDClusterProducer::produce(edm::Event& e, const edm::EventSetup& es)
+{
+  // Step A.1: get input data
+  edm::Handle< FTLRecHitCollection >  inputBarrel;
+  edm::Handle< FTLRecHitCollection >  inputEndcap;
+  e.getByToken(btlHits_, inputBarrel);
+  e.getByToken(etlHits_, inputEndcap);
+      
+  // Step A.2: get event setup
+  edm::ESHandle<MTDGeometry> geom;
+  es.get<MTDDigiGeometryRecord>().get(geom);
+  geom_ = geom.product();
+
+  edm::ESHandle<MTDTopology> mtdTopo;
+  es.get<MTDTopologyRcd>().get(mtdTopo);
+  topo_ = mtdTopo.product();
```

Comment:
If `geom_` and `topo_` would be given to `run()` as arguments, this module could become global.